crates/aptos-jwk-consensus/src/jwk_manager_per_key.rs

```diff
@@ -0,0 +1,442 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    jwk_observer::JWKObserver,
+    mode::per_key::PerKeyMode,
+    network::IncomingRpcRequest,
+    types::{
+        ConsensusState, JWKConsensusMsg, ObservedKeyLevelUpdate, ObservedKeyLevelUpdateRequest,
+        ObservedUpdate, ObservedUpdateResponse, QuorumCertProcessGuard,
+    },
+    update_certifier::{TUpdateCertifier, UpdateCertifier},
+    TConsensusManager,
+};
+use anyhow::{anyhow, bail, Context, Result};
+use aptos_channels::{aptos_channel, message_queues::QueueStyle};
+use aptos_crypto::{bls12381::PrivateKey, SigningKey};
+use aptos_logger::{debug, error, info, warn};
+use aptos_reliable_broadcast::ReliableBroadcast;
+use aptos_types::{
+    account_address::AccountAddress,
+    epoch_state::EpochState,
+    jwks::{
+        jwk::JWK, AllProvidersJWKs, Issuer, KeyLevelUpdate, OIDCProvider, ObservedJWKs,
+        ObservedJWKsUpdated, ProviderJWKsIndexed, QuorumCertifiedUpdate, SupportedOIDCProviders,
+        KID,
+    },
+    validator_txn::{Topic, ValidatorTransaction},
+};
+use aptos_validator_transaction_pool::VTxnPoolState;
+use futures_channel::oneshot;
+use futures_util::{future::join_all, FutureExt, StreamExt};
+use std::{
+    collections::{HashMap, HashSet},
+    sync::Arc,
+    time::Duration,
+};
+use tokio_retry::strategy::ExponentialBackoff;
+
+/// `JWKManager` executes per-issuer JWK consensus sessions
+/// and updates validator txn pool with quorum-certified JWK updates.
+pub struct KeyLevelConsensusManager {
+    /// Some useful metadata.
+    my_addr: AccountAddress,
+    epoch_state: Arc<EpochState>,
+
+    /// Used to sign JWK observations before sharing them with peers.
+    consensus_key: Arc<PrivateKey>,
+
+    /// The sub-process that collects JWK updates from peers and aggregate them into a quorum-certified JWK update.
+    update_certifier: Arc<dyn TUpdateCertifier<PerKeyMode>>,
+
+    /// When a quorum-certified JWK update is available, use this to put it into the validator transaction pool.
+    vtxn_pool: VTxnPoolState,
+
+    onchain_jwks: HashMap<Issuer, ProviderJWKsIndexed>,
+    /// The JWK consensus states of all (issuer,kid)s.
+    states_by_key: HashMap<(Issuer, KID), ConsensusState<ObservedKeyLevelUpdate>>,
+
+    /// Whether a CLOSE command has been received.
+    stopped: bool,
+
+    qc_update_tx: aptos_channel::Sender<(Issuer, KID), QuorumCertifiedUpdate>,
+    qc_update_rx: aptos_channel::Receiver<(Issuer, KID), QuorumCertifiedUpdate>,
+    jwk_observers: Vec<JWKObserver>,
+}
+
+impl KeyLevelConsensusManager {
+    pub fn new(
+        consensus_key: Arc<PrivateKey>,
+        my_addr: AccountAddress,
+        epoch_state: Arc<EpochState>,
+        rb: ReliableBroadcast<JWKConsensusMsg, ExponentialBackoff>,
+        vtxn_pool: VTxnPoolState,
+    ) -> Self {
+        let update_certifier = Arc::new(UpdateCertifier::new(rb));
+
+        let (qc_update_tx, qc_update_rx) = aptos_channel::new(QueueStyle::KLAST, 1, None);
```

Comment:
is it possible to have multiple qc that overrides each other with size 1?

Reply:
while it's possible that QC2 is pushed before QC1 is popped, it shouldn't matter which QC is proposed (or even whether some QC gets proposed or not)? As long as the provider has eventual consistency in a KID, we will have it too.